### DISCUSSION TITLE
Feature Request: Python: Support Arrow PyCapsule Interface & remove `pyarrow` dependency

### DISCUSSION THREAD

**Discussion by kylebarron** [duckdb/duckdb#10716] [Ideas]
👋 The Arrow project recently created the [Arrow PyCapsule Interface](https://arrow.apache.org/docs/format/CDataInterface/PyCapsuleInterface.html), a new protocol for sharing Arrow data in Python. It enables safe Arrow data interchange **without requiring the use of pyarrow**. Any library implementing this protocol can exchange data via PyCapsules (safe wrappers of C pointers), where data producer and consumer don't need prior knowledge of each other.

PyArrow is an extremely large dependency (the unzipped manylinux wheel is 130MB, not including another 69MB for its required numpy dependency), which makes it hard to use DuckDB + Arrow in some environments.

There are a [growing number of Python libraries that support this protocol](https://github.com/apache/arrow/issues/39195#issuecomment-2245718008):

- Polars v1.3+
- pyarrow v14+
- Pandas v2.2+ (export only) https://github.com/pandas-dev/pandas/pull/56587
- [ibis](https://github.com/ibis-project/ibis) https://github.com/ibis-project/ibis/pull/9143, https://github.com/ibis-project/ibis/pull/9665
- [nanoarrow](https://github.com/apache/arrow-nanoarrow)
- datafusion https://github.com/apache/datafusion-python/pull/825, https://github.com/apache/datafusion-python/issues/752
- quak: https://github.com/manzt/quak/pull/23
- a bunch of niche libraries: [pyogrio](https://github.com/geopandas/pyogrio), [gdal](https://github.com/OSGeo/gdal), [lonboard](https://github.com/developmentseed/lonboard), [arro3](https://github.com/kylebarron/arro3)

So for example once DuckDB implements this protocol, DuckDB and Polars will have bidirectional exchange of Arrow data **without pyarrow**.

More libraries exist where implementation is in progress or interested:

- vegafusion: https://github.com/vega/vegafusion/pull/501
- fastexcel: https://github.com/ToucanToco/fastexcel/issues/268
- daft: https://github.com/Eventual-Inc/Daft/issues/2504
- lance: https://github.com/lancedb/lance/issues/2630
- lightgbm: https://github.com/microsoft/LightGBM/issues/6204#issuecomment-2256015538

### Data export

DuckDB can implement data export `__arrow_c_stream__` on `DuckDBPyRelation`, to export a C stream of Arrow data to Python.

### Data import

DuckDB currently looks only for Pandas, Polars, or PyArrow objects when making replacement scans. It should also look for objects that have an `__arrow_c_stream__` method (and potentially also those that have an `__arrow_c_array__` method, which export a single Arrow RecordBatch).

### User-defined functions

DuckDB [allows for using Arrow for user-defined functions](https://duckdb.org/docs/api/python/function.html), but it is currently tied to pyarrow. Supporting the PyCapsule Interface for UDFs would allow, e.g. using Polars for a DuckDB user-defined function, without pyarrow as an intermediary.

**Comment by kylebarron** [comment]
I reworked the argument in the original post to show that there are really broad benefits here in enabling DuckDB to interoperate with the growing Arrow ecosystem in Python, without requiring pyarrow.

I also wanted to tag @Alex-Monahan because we talked about this at the SciPy conference a couple weeks ago.

**Comment by kylebarron** [comment]
Another benefit of supporting the PyCapsule Interface is that it lets DuckDB ask the data producer to cast the exported data to a specific set of supported Arrow types. In https://github.com/manzt/quak/issues/41 we learned that DuckDB appears not to support Arrow string view types as input (at least as of the time of writing). 

With the [`requested_schema` parameter](https://arrow.apache.org/docs/format/CDataInterface/PyCapsuleInterface.html#arrow_c_stream__) of the interface, DuckDB can [request](https://arrow.apache.org/docs/format/CDataInterface/PyCapsuleInterface.html#schema-requests) that the producer cast its exported data to data types that DuckDB knows it can support.

**Reply by Tishj** [reply]
DuckDB will not be in control of capsule creation (such as calling `__arrow_c_stream__` on a pa.Table)

When this is implemented DuckDB will be able to consume a pycapsule object.
How that object is created is up to the user.

String view should be supported as evidenced by [this test](https://github.com/duckdb/duckdb/blob/main/tools/pythonpkg/tests/fast/arrow/test_arrow_string_view.py) added by <https://github.com/duckdb/duckdb/pull/10481>, I'm not aware of that limitation, did you raise this issue here?

**Reply by Tishj** [reply]
The interface documentation makes it clear that this is *not* one of the goals:

> Standardize what public APIs should be used for import. This is left up to individual libraries.

Which makes me very confused why there is a "requested schema" parameter.

This has to be a conversation between consumer and producer, that conversation starts with "I am a producer, without modification this is the schema I will be providing".

Not only does this now start the conversation from the other side: "I am a consumer, I request this schema" it also breaks the communication line from the producer<->consumer, entering the user as a middleman

**Reply by Tishj** [reply]
Hmm actually, I think perhaps I misread through the (admittedly way too convoluted / padded) interface docs

> Standardize what public APIs should be used for import. This is left up to individual libraries.

Only means the consumer side, it *does* specify that `__arrow_c_stream__` should be used to create a PyCapsule

I was under the impression they were just listing changes made to PyArrow at first

**Reply by kylebarron** [reply]
> DuckDB will not be in control of capsule creation (such as calling `__arrow_c_stream__` on a 
pa.Table)

Sorry, I'm not sure I understand what you're saying here. Are you saying DuckDB doesn't _want_ to look for an `__arrow_c_stream__` method or that it can't for some reason?

The interface tries to keep raw capsules out of view of Python users, so it's intended that `__arrow_c_stream__` be called from other Arrow-based libraries, rather than being called by the end user.

> String view should be supported as evidenced by [this test](https://github.com/duckdb/duckdb/blob/main/tools/pythonpkg/tests/fast/arrow/test_arrow_string_view.py) added by #10481, I'm not aware of that limitation, did you raise this issue here?

I can create a separate bug report for this. This fails: (see issue https://github.com/duckdb/duckdb/issues/13424)

```py
import polars as pl
import duckdb
import pyarrow as pa

df = pl.DataFrame({"a": ["a", "b", "c"]})
table = pa.table(df)
duckdb.from_arrow(table)
```

```
---------------------------------------------------------------------------
NotImplementedException                   Traceback (most recent call last)
File /Users/kyle/github/kylebarron/arro3/tests/core/test_ffi.py:[1](https://file+.vscode-resource.vscode-cdn.net/Users/kyle/github/kylebarron/arro3/tests/core/test_ffi.py:1)
----> 1 duckdb.from_arrow(table)

File ~/github/kylebarron/arro3/.venv/lib/python3.11/site-packages/duckdb/__init__.py:505, in from_arrow(arrow_object, **kwargs)
    [503](https://file+.vscode-resource.vscode-cdn.net/Users/kyle/github/kylebarron/arro3/~/github/kylebarron/arro3/.venv/lib/python3.11/site-packages/duckdb/__init__.py:503) else:
    [504](https://file+.vscode-resource.vscode-cdn.net/Users/kyle/github/kylebarron/arro3/~/github/kylebarron/arro3/.venv/lib/python3.11/site-packages/duckdb/__init__.py:504)     conn = duckdb.connect(":default:")
--> [505](https://file+.vscode-resource.vscode-cdn.net/Users/kyle/github/kylebarron/arro3/~/github/kylebarron/arro3/.venv/lib/python3.11/site-packages/duckdb/__init__.py:505) return conn.from_arrow(arrow_object, **kwargs)

NotImplementedException: Not implemented Error: Unsupported Internal Arrow Type vu
```

With polars 1.4.1, duckdb 1.0.0, pyarrow 17.0.0. 


> The interface documentation makes it clear that this is _not_ one of the goals:
> 
> > Standardize what public APIs should be used for import. This is left up to individual libraries.

I think that means the standard has no bearing on what DuckDB calls its `from_arrow` method. But whatever DuckDB's primary Arrow API is, ideally it should look for objects exporting the PyCapsule Interface.

> This has to be a conversation between consumer and producer, that conversation starts with "I am a producer, without modification this is the schema I will be providing".

Yes. In https://github.com/apache/arrow/issues/39689 the consensus seems to be that objects exporting `__arrow_c_stream__` should **also** export `__arrow_c_schema__` so that data consumers can

1. see the Arrow schema of the producer's data
2. decide whether to ask for data with a different Arrow schema
3. call `__arrow_c_stream__` with that other Arrow schema if desired

**Reply by WillAyd** [reply]
Hi - Kyle just introduced me to this discussion, so figured I'd chime in with things that might be helpful.

Regarding the string view issue, DuckDB just vendors a copy of nanoarrow, right? If so, I believe that's the source of the issue - nanoarrow 0.5.0 did not implement support for that, although I think it might make its way into the next release (see also https://github.com/apache/arrow-nanoarrow/pull/596)

I've also been recently providing video walk throughs and a github repo showing developers how to use nanoarrow with C/C++ extensions to consume data via the PyCapsule interface. My content uses nanobind, and I see DuckDB uses pybind11, so I don't think should be that drastically different:

https://github.com/WillAyd/bearly/pull/1/files

The consumption part is relatively easy - you can just inspect the Python object for any of the required dunders. On the export side, I do think the specification is vague. If you don't know who the consumer is for data you produce, then I'm not sure what object you actually can provide outside of a capsule. If DuckDB already has or develops an object that can wrap the exported capsule in an object implementing the PyArrow interface dunders then that may be a non-issue; if that doesn't exist then I'm not sure what the right solution is, but it's possible I am overlooking something easy

**Reply by kylebarron** [reply]
> Regarding the string view issue

The issue above was reported in https://github.com/duckdb/duckdb/issues/13424 and is fixed on latest main.

> I'm not sure what object you actually can provide outside of a capsule. 

`DuckDBPyRelation` already has a [`fetch_arrow_reader` method](https://duckdb.org/docs/api/python/reference/#duckdb.DuckDBPyRelation.fetch_arrow_reader), which returns a `pyarrow.RecordBatchReader`. Therefore, I'd argue that `DuckDBPyRelation` should also expose an `__arrow_c_stream__` dunder method. The same code underlying `fetch_arrow_reader` can be reused to create the Arrow C Stream capsule exported via `__arrow_c_stream__`.

**Reply by Tishj** [reply]
DuckDB does not vendor a copy of nanoarrow

**Reply by Tishj** [reply]
> Therefore, I'd argue that `DuckDBPyRelation` should also expose an `__arrow_c_stream__` dunder method.


https://github.com/duckdb/duckdb/pull/13418

**Reply by jorisvandenbossche** [reply]
Thanks a lot @Tishj for implementing this! (and for the other PRs supporting the consume side, https://github.com/duckdb/duckdb/pull/13386and https://github.com/duckdb/duckdb/pull/13425)

I took a quick look at them, and left a comment at https://github.com/duckdb/duckdb/pull/13418#discussion_r1746662887 (I didn't test this in practice, but my guess is that this would require fixing to allow pyarrow to consume a duckdb pyrelation)

**Comment by Tishj** [comment]
I'm closing this discussion as the relevant functionality has been added and will be part of the next release (1.1.0)
